### PR TITLE
Add Cash App Pay for v5

### PR DESCRIPTION
**Pull request by OscarSpruit**
## Description
Add the Cash App Pay payment method for v5.

## Checklist 
- [x] Code is unit tested
- [x] Changes are tested manually
- [x] Add relevant labels to PR  

COAND-634

### CODE REVIEW

File: .idea/copyright/profiles_settings.xml [hunk 1]
```diff
@@ -1,7 +1,3 @@
 <component name="CopyrightManager">
-  <settings default="Adyen MIT">
-    <module2copyright>
-      <element module="Project Files" copyright="Adyen MIT" />
-    </module2copyright>
-  </settings>
-</component>
+  <settings default="Adyen MIT" />
+</component>
```
Comment (jreij) [inline .idea/copyright/profiles_settings.xml:3]: Is this change intended?

File: example-app/src/main/java/com/adyen/checkout/example/ui/configuration/CheckoutConfigurationProvider.kt [hunk 2]
```diff
@@ -63,6 +65,11 @@ internal class CheckoutConfigurationProvider @Inject constructor(
             .setAmount(amount)
             .build()
 
+    private fun getCashAppPayConfiguration(): CashAppPayConfiguration =
+        CashAppPayConfiguration.Builder(shopperLocale, environment, clientKey)
+            .setReturnUrl("adyencheckout://com.adyen.checkout.example")
```
Comment (jreij) [inline example-app/src/main/java/com/adyen/checkout/example/ui/configuration/CheckoutConfigurationProvider.kt:70]: Can we use `CashAppPayComponent.getReturnUrl` instead?

File: cashapppay/src/main/java/com/adyen/checkout/cashapppay/internal/ui/model/CashAppPayComponentParamsMapper.kt [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+        )
+    }
+
+    private fun CashAppPayComponentParams.override(
+        sessionParams: SessionParams?,
+    ): CashAppPayComponentParams {
+        if (sessionParams == null) return this
+        return copy(
+            amount = sessionParams.amount ?: amount,
+            // TODO: Check how this should be overridden
```
Comment (jreij) [inline cashapppay/src/main/java/com/adyen/checkout/cashapppay/internal/ui/model/CashAppPayComponentParamsMapper.kt:72]: Do not forget this

File: cashapppay/src/main/java/com/adyen/checkout/cashapppay/internal/ui/model/CashAppPayComponentParamsMapper.kt [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+            showStorePaymentField = sessionParams.enableStoreDetails ?: showStorePaymentField,
+            // TODO: Check if returnUrl can be overridden
```
Comment (jreij) [inline cashapppay/src/main/java/com/adyen/checkout/cashapppay/internal/ui/model/CashAppPayComponentParamsMapper.kt:74]: Do not forget this

File: cashapppay/src/main/java/com/adyen/checkout/cashapppay/internal/ui/model/CashAppPayComponentParamsMapper.kt [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+
+    private fun CashAppPayConfiguration.mapToParamsInternal() = CashAppPayComponentParams(
+        isSubmitButtonVisible = isSubmitButtonVisible ?: true,
+        shopperLocale = shopperLocale,
+        environment = environment,
+        clientKey = clientKey,
+        isAnalyticsEnabled = isAnalyticsEnabled ?: true,
+        isCreatedByDropIn = false,
+        amount = amount,
+        cashAppPayEnvironment = cashAppPayEnvironment ?: getDefaultCashAppPayEnvironment(this),
```
Comment (jreij) [inline cashapppay/src/main/java/com/adyen/checkout/cashapppay/internal/ui/model/CashAppPayComponentParamsMapper.kt:38]: Nitpick: check out `getGooglePayEnvironment` in `GooglePayComponentParamsMapper` in case you find it nicer

File: cashapppay/src/main/java/com/adyen/checkout/cashapppay/CashAppPayComponent.kt [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+/**
+ * A [PaymentComponent] that supports the [PaymentMethodTypes.CASH_APP_PAY] payment method.
+ */
+class CashAppPayComponent internal constructor(
+    private val cashAppPayDelegate: CashAppPayDelegate,
+    private val genericActionDelegate: GenericActionDelegate,
+    private val actionHandlingComponent: DefaultActionHandlingComponent,
+    internal val componentEventHandler: ComponentEventHandler<CashAppPayComponentState>,
+) : ViewModel(),
+    PaymentComponent,
```
Comment (jreij) [inline cashapppay/src/main/java/com/adyen/checkout/cashapppay/CashAppPayComponent.kt:43]: Should this be a `ButtonComponent` as well?

File: cashapppay/src/main/java/com/adyen/checkout/cashapppay/internal/ui/model/CashAppPayParams.kt [hunk 7]
```diff
@@ -0,0 +1,15 @@
...
+ * Copyright (c) 2023 Adyen N.V.
+ *
+ * This file is open source and available under the MIT license. See the LICENSE file for more info.
+ *
+ * Created by oscars on 27/6/2023.
+ */
+
+package com.adyen.checkout.cashapppay.internal.ui.model
+
+internal data class CashAppPayParams(
```
Comment (jreij) [inline cashapppay/src/main/java/com/adyen/checkout/cashapppay/internal/ui/model/CashAppPayParams.kt:11]: Can we remove this class and move its data to `CashAppPayComponentParams`?

File: cashapppay/src/main/java/com/adyen/checkout/cashapppay/internal/ui/DefaultCashAppPayDelegate.kt [hunk 8]
```diff
@@ -0,0 +1,326 @@
...
+            ?: throw ComponentException(
+                "Cannot launch Cash App Pay, set the returnUrl in your CashAppPayConfiguration.Builder"
+            ),
+    )
+
+    private fun initCashAppPay(cashAppPayParams: CashAppPayParams): CashAppPay {
+        return if (componentParams.cashAppPayEnvironment == CashAppPayEnvironment.SANDBOX) {
+            CashAppPayFactory.createSandbox(cashAppPayParams.clientId)
+        } else {
+            CashAppPayFactory.createSandbox(cashAppPayParams.clientId)
```
Comment (jreij) [inline cashapppay/src/main/java/com/adyen/checkout/cashapppay/internal/ui/DefaultCashAppPayDelegate.kt:112]: Should be `.create`

**Review by jreij — CHANGES_REQUESTED** [review]

**Comment by jreij** [comment]
Component should auto submit and not show any button when `setShowStorePaymentField(false)` is called

File: cashapppay/src/main/java/com/adyen/checkout/cashapppay/internal/ui/model/CashAppPayComponentParamsMapper.kt [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+    ) = CashAppPayComponentParams(
+        isSubmitButtonVisible = isSubmitButtonVisible ?: true,
+        shopperLocale = shopperLocale,
+        environment = environment,
+        clientKey = clientKey,
+        isAnalyticsEnabled = isAnalyticsEnabled ?: true,
+        isCreatedByDropIn = false,
+        amount = amount,
+        cashAppPayEnvironment = getDefaultCashAppPayEnvironment(),
+        returnUrl = returnUrl ?: throw ComponentException(
```
Comment (jreij) [inline cashapppay/src/main/java/com/adyen/checkout/cashapppay/internal/ui/model/CashAppPayComponentParamsMapper.kt:63]: Would this crash the component even if the `returnUrl` is set in `/sessions`?

File: cashapppay/src/main/java/com/adyen/checkout/cashapppay/internal/ui/DefaultCashAppPayDelegate.kt [hunk 10]
```diff
@@ -0,0 +1,316 @@
...
+
+        return CashAppPayComponentState(
+            data = paymentComponentData,
+            isInputValid = outputData.isValid,
+            isReady = true
+        )
+    }
+
+    override fun onSubmit() {
+        initiatePayment()
```
Comment (jreij) [inline cashapppay/src/main/java/com/adyen/checkout/cashapppay/internal/ui/DefaultCashAppPayDelegate.kt:187]: Can we add a check `if (isConfirmationRequired())` to make sure this is only triggered when the switch is visible?

File: components-core/src/main/java/com/adyen/checkout/components/core/internal/ui/model/SessionParams.kt [hunk 11]
```diff
@@ -16,4 +16,5 @@ data class SessionParams(
     val enableStoreDetails: Boolean?,
     val installmentOptions: Map<String, SessionInstallmentOptionsParams?>?,
     val amount: Amount?,
+    val returnUrl: String,
```
Comment (jreij) [inline components-core/src/main/java/com/adyen/checkout/components/core/internal/ui/model/SessionParams.kt:19]: Could this be null?

Comment (OscarSpruit) [reply components-core/src/main/java/com/adyen/checkout/components/core/internal/ui/model/SessionParams.kt:19]: No, because in `SessionSetupResponse` it is also not nullable

File: cashapppay/src/main/java/com/adyen/checkout/cashapppay/internal/ui/model/CashAppPayComponentParamsMapper.kt (hunk 9 above)
Comment (OscarSpruit) [reply cashapppay/src/main/java/com/adyen/checkout/cashapppay/internal/ui/model/CashAppPayComponentParamsMapper.kt:63]: Good catch

File: cashapppay/src/main/java/com/adyen/checkout/cashapppay/internal/ui/model/CashAppPayComponentParamsMapper.kt [hunk 12]
```diff
@@ -0,0 +1,113 @@
...
+        .override(overrideComponentParams)
+        .override(sessionParams ?: overrideSessionParams)
+
+    fun mapToParams(
+        configuration: CashAppPayConfiguration,
+        sessionParams: SessionParams?,
+        @Suppress("UNUSED_PARAMETER") paymentMethod: StoredPaymentMethod,
+    ): CashAppPayComponentParams = configuration
+        // clientId and scopeId are not needed in the stored flow.
+        .mapToParamsInternal(null, null, configuration.getReturnUrlOrThrow(sessionParams ?: overrideSessionParams))
```
Comment (jreij) [inline cashapppay/src/main/java/com/adyen/checkout/cashapppay/internal/ui/model/CashAppPayComponentParamsMapper.kt:48]: Hmm this overriding logic is done outside of the `override` method right? Can we change that somehow?

**Review by jreij — APPROVED** [review]
